### PR TITLE
chore: use build tags to select plugins

### PR DESCRIPTION
**Pull request by reimda**
Before this PR, Telegraf's plugins are included in the build if they are imported in a file called all.go. (example plugins/inputs/all/all.go)

This PR uses golang build tags to select which plugins to include in the build. It removes all.go and replaces it with multiple files. Those files have build tags that allow them to be selected or not. It also adds a way to set build tags when building with the makefile. This allows us to make custom builds with a different set of plugins without rewriting all.go files.

The build tags are passed through the makefile by using an environment variable `BUILDTAGS`. If this isn't set, all plugins are included in the build as before.

```bash
$ make telegraf
go build -ldflags " -X main.commit=f56bc051 -X main.branch=buildtags -X main.goos=linux -X main.goarch=amd64 -X main.version=1.24.0-f56bc051" -tags "" ./cmd/telegraf
$ ./telegraf --input-list|wc -l # this is the same number of plugins as before this PR
222
$ ls -lh telegraf # same size as before this PR
-rwxr-xr-x 1 reim reim 197M Aug  9 16:06 telegraf
```

To make a custom build with only the zfs and zookeeper input plugins, set `BUILDTAGS="inputs.zookeeper inputs.zfs"` when running make.

```bash
$ BUILDTAGS="custom inputs.zookeeper inputs.zfs" make telegraf
go build -ldflags " -X main.commit=f56bc051 -X main.branch=buildtags -X main.goos=linux -X main.goarch=amd64 -X main.version=1.24.0-f56bc051" -tags "custom inputs.zookeeper inputs.zfs" ./cmd/telegraf
reim@debian:~/go/src/github.com/influxdata/telegraf$ ./telegraf --input-list # only includes the inputs we asked for
Available Input Plugins:
  snmp
  zfs
  zookeeper
$ ls -lh telegraf # smaller size
-rwxr-xr-x 1 reim reim 105M Aug  9 16:20 telegraf
```
In the previous example, snmp is also included because it is used by one of the processor plugins.

Using build tags like this gives us a way to name subsets of plugins. For example, to include a few "core" plugins, use `BUILDTAGS="core"`. (I manually added the "core" build tag to a few of the files made by the shell script.)

```bash
$ grep -l core plugins/inputs/all/*.go # these are the core plugins
plugins/inputs/all/cpu.go
plugins/inputs/all/disk.go
plugins/inputs/all/diskio.go
plugins/inputs/all/kernel.go
plugins/inputs/all/mem.go
plugins/inputs/all/processes.go
plugins/inputs/all/swap.go
plugins/inputs/all/system.go
$ BUILDTAGS="custom core" make telegraf
go build -ldflags " -X main.commit=f56bc051 -X main.branch=buildtags -X main.goos=linux -X main.goarch=amd64 -X main.version=1.24.0-f56bc051" -tags "custom core" ./cmd/telegraf
reim@debian:~/go/src/github.com/influxdata/telegraf$ ./telegraf --input-list
Available Input Plugins:
  cpu
  disk
  diskio
  io
  kernel
  linux_sysctl_fs
  mem
  processes
  snmp
  swap
  system
$ ls -lh telegraf 
-rwxr-xr-x 1 reim reim 105M Aug  9 16:19 telegraf
```
In the previous example, linux_sysctl_fs plugin is also included because the processes input plugin uses it.

#11524 adds a tool to make custom telegraf builds that works by generating all.go files. Since it generates code, it has to worry about cleaning up after the build completes so the generated code won't accidentally be committed. If we used build tags instead, there would be no need for code generation or cleanup.

### CODE REVIEW

**Comment by reimda** [comment]
@srebhan what do you think of us doing something like this in https://github.com/influxdata/telegraf/pull/11524 instead of generating the all.go files?

**Comment by srebhan** [comment]
@reimda I also thought about this approach and came up with the following points to consider:
- We need to have a way to make sure that _only_ build has the ability to exclude plugins. The linter, checks and all other makefile targets should _always_ run for _all_ plugins. I guess that's easily done in the Makefile...
- We have to have an "import" file for every plugin plus an empty one that is always build (if you choose no plugin in that category). That's a review/CI task we have to be aware of for new plugins, but we can easily generate the "all" files with a tool (minify has all the information anyway).
- The most important one: This is hard to use. 
  - If you have more than a handful of plugins you want to select, the command-line becomes crowded.
  - There is no "built-in" way to store the configuration you want to build, so the user has to manually store the command somewhere or write a script or anything.
  - For the user it needs manual work to convert the config files to tags in the command line. We would need a tool again...
  - Typos in the tags will go unnoticed until the user runs the binary to see what's in! Think of `input.modbus` or `outputs.influxdbv2`...

However, I also see the benefits, like 
- this being built-in without additional code
- no messing around with files and no "build state"
- the ability to select plugins by platform from the `all` files, which currently is hard to do
- being able to generate those files to avoid issues
- full and easy control of when things are customized...

For the usage reasons and the amount of change to the `all.go` files I went the current minify route...
This being said, we should discuss this in the team and come up with a way to "fix" the usage issues. Maybe `minify` could just be a tool to construct the tags and runs the make command, doing the checks etc for you. Then, this would be the ideal solution I think!

**Comment by reimda** [comment]
> This is hard to use

> Maybe `minify` could just be a tool to construct the tags and runs the make command, doing the checks etc for you.

I agree that it's not easy enough to use by itself so I was also thinking it would be part of the tool aka 'minify'. The tool would figure out which plugins are needed given a telegraf.conf, generate a list of tags, then set the environment variable and run make.

**Comment by srebhan** [comment]
@reimda nice! Some comments:

1. I used the following script [make_files.sh.txt](https://github.com/influxdata/telegraf/files/9307316/make_files.sh.txt) to generate the files avoiding the linter errors regarding blank imports.
2. The `influx` parser should contain
```golang
//go:build !custom || parsers || parsers.influx

package all

import (
        _ "github.com/influxdata/telegraf/plugins/parsers/influx"                 // register plugin
        _ "github.com/influxdata/telegraf/plugins/parsers/influx/influx_upstream" // register plugin
)
```
and `plugins/parsers/all/influx_upstream.go` should be removed, as the two are switchable by a flag and need to included both for `data_format="influx"`.
3. `plugins/processors/all/ec2.go` should actually be `plugins/processors/all/aws_ec2.go` and the build tag should be `processors.aws_ec2` as this is the registered name.
4. Can you please remove the `all.go-old` files from the PR?
5. The Makefile introduces the `BUILDTAGS` at too many locations. I think we only need to add it to the `build` target.

**Comment by reimda** [comment]
> 5. The Makefile introduces the `BUILDTAGS` at too many locations. I think we only need to add it to the `build` target.

I was thinking `-tags $(BUILDTAGS)` would be useful for all the targets that call build, install, and test, but it's only really needed now in the `build:` tag. I'm ok with removing them until we need them.

* install: I'm not sure `go-install:` is even in use any more.
* test: Using tags optionally reduces the number of tests which isn't what we want most of the time.
* build: we'll have to add it to the `$(buildbin)` target if we try to make packages containing a custom build. We can figure that out later.

File: plugins/inputs/all/make_files.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    IFS= cat > "$plugin.go" <<EOF
+//go:build !custom || inputs || inputs.$plugin
+
+package all
+
+import (
+$line
+)
+EOF
+done < all.go-old
```
Comment (powersj) [inline plugins/inputs/all/make_files.sh:34]: do we need this file checked in after doing this?

**Review by powersj — COMMENTED** [review]
Thanks for putting this together. One question in line, and I would ask for two additional updates:

1) Please update the developer's documents with what needs to happen when adding a new plugin. (e.g. [inputs](https://github.com/influxdata/telegraf/blob/master/docs/INPUTS.md), [outputs](https://github.com/influxdata/telegraf/blob/master/docs/OUTPUTS.md), [aggregators](https://github.com/influxdata/telegraf/blob/master/docs/AGGREGATORS.md), [processors](https://github.com/influxdata/telegraf/blob/master/docs/PROCESSORS.md))

2. As you have mentioned in previous chats, users who know what they are doing know how to remove plugins from the all.go file. With those entries gone, can we have a short doc, which essentially has part of your PR description in it, added to the docs?

Thanks!

File: plugins/inputs/all/make_files.sh (hunk 1 above)
Comment (reimda) [reply plugins/inputs/all/make_files.sh]: There's no need for it once the files are generated. It's removed now.

**Review by sspaink — APPROVED** [review]
Awesome work, this looks like a great solution.

**Review by powersj — APPROVED** [review]
Thank you for the doc updates!

**Pull request merged**